### PR TITLE
[21][record patterns] type inference with wildcard

### PR DESCRIPTION
**Pull request by stephan-herrmann**
Test case for  #1336

### CODE REVIEW

**Comment by stephan-herrmann** [comment]
In the second commit I cheekily implemented what I guess to be missing from JLS 18.5.5.

This is a long shot needing a confirmation from a spec author or javac developer. For the simple test from #1336 it suffices, and it "suffices" to break test `testRecordPatternTypeInference_010()` which no fails to generate a necessary instanceof test, resulting in` java.lang.ClassCastException: class X$$Lambda/0x00000008000c0c00 cannot be cast to class java.lang.String`

Also the tweak in `shouldInitiateRecordTypeInference()` is likely illegal.

**Comment by stephan-herrmann** [comment]
@mpalat let me know if you want me to translate the impl sketch back into spec language that could be used for discussion with the spec author.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java [hunk 1]
```diff
@@ -2005,6 +2010,25 @@ public ReferenceBinding inferRecordPatternParameterization(
...
+			Pattern[] patterns = recordPattern.patterns;
+			RecordComponentBinding[] components = ((ReferenceBinding) recordPattern.resolvedType.original()).components();
+			List<ConstraintFormula> extendedConstraints = new ArrayList<>();
+			if (this.initialConstraints != null)
+				extendedConstraints.addAll(Arrays.asList(this.initialConstraints));
+			for (int i = 0; i < patterns.length; i++) {
+				Pattern pattern = patterns[i];
+				if (pattern instanceof TypePattern && pattern.resolvedType != null && pattern.resolvedType.isValidBinding()) {
+					TypeBinding compTypeSubst = substitute(components[i].type);
+					if (!compTypeSubst.isProperType(true))
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java:2024]: this check could be wrong, perhaps the constraint needs to be created always, to help detecting incompatible situations.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java [hunk 2]
```diff
@@ -2005,6 +2010,25 @@ public ReferenceBinding inferRecordPatternParameterization(
 		return getRecordPatternTypeFromUpwardsProjection(typeBinding, alphas, solution);
 	}
 
+	public void addConstraintsFromNestedPatterns(RecordPattern recordPattern) {
+		if (recordPattern.patterns != null) {
+			Pattern[] patterns = recordPattern.patterns;
+			RecordComponentBinding[] components = ((ReferenceBinding) recordPattern.resolvedType.original()).components();
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java:2016]: calling `original()` probably is cheating. I guess we need to peel raw types but not parameterized types.

**Comment by mpalat** [comment]
> @mpalat let me know if you want me to translate the impl sketch back into spec language that could be used for discussion with the spec author.

Yes, I would definitely require .. however, before that , I just want to confirm that it is not a bug from our side - ref https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1336#issuecomment-1703745394

**Comment by stephan-herrmann** [comment]
Test case was copied to #1341--  the rest (second commit) is probably incorrect.

**Pull request closed without merging**